[POC] Provisioner for SBOM (#13171)
hashicorp/packer: website/content/docs/provisioners/hcp-sbom.mdx

---
description: |
  The `hcp-sbom` Packer provisioner downloads an SBOM file from the guest VM and
  sends it to HCP Packer when the build is done.
page_title: HCP SBOM - Provisioners
---

<BadgesHeader>
  <PluginBadge type="official"/>
</BadgesHeader>

# HCP SBOM Provisioner

Type: `hcp-sbom`

The `hcp-sbom` Packer provisioner downloads an SBOM file from the guest machine
and sends it to HCP Packer when the build is complete (only if the template is
HCP-enabled). The SBOM file is automatically removed at the end of the process.
If you want to retain a copy of the SBOM file, you can specify the
`destination` option in the provisioner.

Currently, we support `CycloneDX` and `SPDX` SBOM formats in `JSON`.

> 

Author: TBA: Add more details about max number of files allowed to download, and if we are going to add the file name field!

---
description: |

>   The `hcp-sbom` Packer provisioner downloads the SBOM file from remote and
>   sends to HCP Packer when build is done.

Reviewer: ```suggestion
  The `hcp-sbom` Packer provisioner downloads an SBOM file from the guest VM and
  sends it to HCP Packer when the build is done.
```

page_title: HCP SBOM - Provisioners
---

<BadgesHeader>
  <PluginBadge type="official"/>
</BadgesHeader>

# HCP SBOM Provisioner

Type: `hcp-sbom`

> The `hcp-sbom` Packer provisioner downloads the SBOM file from the remote machine

Reviewer: ```suggestion
The `hcp-sbom` Packer provisioner downloads an SBOM file from the guest machine
```

and sends it to HCP Packer when the build is complete (only if the template is
HCP-enabled). The SBOM file is automatically removed at the end of the process.

> If the user wants to retain a copy of the SBOM file, they should specify the
> `destination` field in the provisioner.

Reviewer: ```suggestion
If you want to retain a copy of the SBOM file, you can specify the
`destination` option in the provisioner.
```

Currently, we support `CycloneDX` and `SPDX` SBOM formats in `JSON`.

> TBA: Add more details about max number of files allowed to download, and

Reviewer: Is that a TODO for us? We should not publish that I'd think

Author: We are not going to publish this but just added it here until we finish the adding other stuff. I am removing this part in my next commit. Will add the PR comment.

Reviewer: Sounds good to me!

  if we are going to add the file name field.

## Basic Example

> <Tabs>

Reviewer: I know that we've used tabs for several components, but recently we've opted to expose that information as sections in the page.
Also I would suggest placing HCL2 first as it's the lingua franca for Packer these days even if JSON is still supported.

      "inline": [
        "apt-get update -y",
        "apt-get install -y curl",
        "bash -c \"$(curl -sSL https://install.mondoo.com/sh)\""
      ]
    },
    {
      "type": "shell",
      "inline": [
        "cnquery sbom --output cyclonedx-json --output-target /tmp/sbom_cyclonedx.json",
        "cnquery sbom --output spdx-json --output-target /tmp/sbom_spdx.json"
      ]
    },
    {
      "type": "hcp-sbom",
      "source": "/tmp/sbom_cyclonedx.json",
      "destination": "./sbom"
    },
    {
      "type": "hcp-sbom",
      "source": "/tmp/sbom_spdx.json",
      "destination": "./sbom/sbom_spdx.json"
    }
  ]
}
```
</Tab>

<Tab heading="HCL2">

```hcl
packer {
  required_plugins {
    docker = {
      version = ">= 1.0.0"
      source  = "github.com/hashicorp/docker"
    }
  }
}

source "docker" "ubuntu" {
  image  = "ubuntu:20.04"
  commit = true
}

build {
  sources = ["source.docker.ubuntu"]

  hcp_packer_registry {
    bucket_name = "test-bucket"
  }

  provisioner "shell" {
    inline = [
      "apt-get update -y",
      "apt-get install -y curl gpg",
      "bash -c \"$(curl -sSL https://install.mondoo.com/sh)\""
    ]
  }

  provisioner "shell" {

>     inline = [

Reviewer: I'd suggest merging both `shell` blocks for the example here, they don't need to be two separate provisioners.
Also I'd suggest keeping the example minimal and only do one of SPDX/CycloneDX.
